lib/Yield/Calls/Slices.cpp

```diff
@@ -0,0 +1,127 @@
+/// \file Slices.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <unordered_map>
+
+#include "llvm/ADT/BreadthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+
+#include "revng/Yield/Calls/Slices.h"
+
+using NodeView = const yield::Graph::Node *;
+
+struct NodeLookupHelper {
+  const MetaAddress &TargetAddress;
+  bool operator()(NodeView Node) { return Node->Address == TargetAddress; }
+};
+
+static yield::Graph::Node *
+nodeHelper(yield::Graph &Graph, const yield::Graph::Node *Source) {
+  auto New = std::make_unique<yield::Graph::Node>(Source->data());
+  return Graph.addNode(std::move(New));
+}
+
+template<bool IsForwards>
+yield::Graph
+preservingSliceImpl(const yield::Graph &Input, const MetaAddress &SlicePoint) {
+  auto Entry = llvm::find_if(Input.nodes(), NodeLookupHelper{ SlicePoint });
+  revng_assert(Entry != Input.nodes().end());
+
+  using NV = std::conditional_t<IsForwards, NodeView, llvm::Inverse<NodeView>>;
+  using INV = std::conditional_t<IsForwards, llvm::Inverse<NodeView>, NodeView>;
+
+  //using Trait = llvm::GraphTraits<NV>;
+  using InverseTrait = llvm::GraphTraits<INV>;
+
+  // Find the rank of each node, such that for any node its rank is equal to
+  // the highest rank among its children plus one.
+  std::unordered_map<const yield::Graph::Node *, size_t> Ranks;
+  llvm::ReversePostOrderTraversal<NodeView, InverseTrait> RPOT(*Entry);
+  for (auto Iterator = RPOT.begin(); Iterator != RPOT.end(); ++Iterator) {
+    uint64_t &CurrentRank = Ranks[*Iterator];
+
+    CurrentRank = 0;
+    for (auto ChildIterator = InverseTrait::child_begin(*Iterator);
+         ChildIterator != InverseTrait::child_end(*Iterator);
+         ++ChildIterator) {
```

Comment:
Why not the following?
```c++
for (auto *Child : llvm::children<INV>(*Iterator))
```

Reply:
(This is the same as the other thread, so I'll close this one)